activation offloading reordering for comp<>comm overlaps

xuanzhang816 opened the pull request:
We introduce two flags for computation <> communication overlap:
* `activation_offload_sink_wait` -- with this turned on, the wait events in forward is sinked to the end of the graph so that the offload ops is not blocking
* `activation_reload_prefetch` -- with this turned on, the reload in the backward is prefetched just enough so that the reload operation is perfectly overlapped with computation.

Continue with the [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4) in the first PR in the stack:

4. AO on separate stream with reorders (i.e., additionally with `activation_offload_sink_wait = True` and `activation_reload_prefetch = True`)

- 7534 MB peak memory 
- 160.86 ms ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="869" height="123" alt="image" src="https://github.com/user-attachments/assets/6aa0c5df-5cc4-4711-8c1a-20f4a9d9145b" />



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #168316
* #167880



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @chenyang78

xuanzhang816 committed:
[wip] activation offloading reordering

[ghstack-poisoned]

xuanzhang816 committed:
Update on "[wip] activation offloading reordering"

We introduce two flags for computation <> communication overlap:
* `activation_offload_sink_wait` -- with this turned on, the wait events in forward is sinked to the end of the graph so that the offload ops is not blocking
* `activation_reload_prefetch` -- with this turned on, the reload in the backward is prefetched just enough so that the reload operation is perfectly overlapped with computation.

Continue with the [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4) in the first PR in the stack:

4. AO on separate stream with reorders (i.e., additionally with `activation_offload_sink_wait = True` and `activation_reload_prefetch = True`)

- 7534 MB peak memory 
- 160.86 ms ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="869" height="123" alt="image" src="https://github.com/user-attachments/assets/6aa0c5df-5cc4-4711-8c1a-20f4a9d9145b" />






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

xuanzhang816 committed:
Update on "activation offloading reordering for comp<>comm overlaps"

We introduce two flags for computation <> communication overlap:
* `activation_offload_sink_wait` -- with this turned on, the wait events in forward is sinked to the end of the graph so that the offload ops is not blocking
* `activation_reload_prefetch` -- with this turned on, the reload in the backward is prefetched just enough so that the reload operation is perfectly overlapped with computation.

Continue with the [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4) in the first PR in the stack:

4. AO on separate stream with reorders (i.e., additionally with `activation_offload_sink_wait = True` and `activation_reload_prefetch = True`)

- 7534 MB peak memory 
- 160.86 ms ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="869" height="123" alt="image" src="https://github.com/user-attachments/assets/6aa0c5df-5cc4-4711-8c1a-20f4a9d9145b" />






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

xuanzhang816 committed:
Update on "activation offloading reordering for comp<>comm overlaps"

We introduce two flags for computation <> communication overlap:
* `activation_offload_sink_wait` -- with this turned on, the wait events in forward is sinked to the end of the graph so that the offload ops is not blocking
* `activation_reload_prefetch` -- with this turned on, the reload in the backward is prefetched just enough so that the reload operation is perfectly overlapped with computation.

Continue with the [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4) in the first PR in the stack:

4. AO on separate stream with reorders (i.e., additionally with `activation_offload_sink_wait = True` and `activation_reload_prefetch = True`)

- 7534 MB peak memory 
- 160.86 ms ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="869" height="123" alt="image" src="https://github.com/user-attachments/assets/6aa0c5df-5cc4-4711-8c1a-20f4a9d9145b" />






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

eellison commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -462,6 +501,261 @@ def put_offload_nodes_on_separate_stream(
     add_backward_reload_stream_ops(bwd_module.graph)
 
 
+def _validate_pattern_nodes(
```
would this be better done with the pattern matcher ?

eellison commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -462,6 +501,261 @@ def put_offload_nodes_on_separate_stream(
+def _estimate_transfer_time_in_ms(transfer_size_bytes: int) -> float:
+    """Estimate transfer time in milliseconds based on size and bandwidth."""
+
+    return transfer_size_bytes / (1024**3) * 1_000 / inductor_config.cpu_gpu_bw
```
potentially this would eventually be standardized in node estimator class..

eellison reviewed (approved):
looks good, just nits. can we update cpu_gpu_bw before landing? On gb200 this is significantly higher than you have here

xuanzhang816 commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -462,6 +501,261 @@ def put_offload_nodes_on_separate_stream(
+    """
+    Identify backward reload patterns in the graph.
+
+    Pattern: fork → wait_stream → device_put → record_event → join → wait_event
```
I see your point. let me add the check!

xuanzhang816 commented on torch/_functorch/_activation_offloading/activation_offloading.py:
```diff
@@ -462,6 +501,261 @@ def put_offload_nodes_on_separate_stream(
+    """
+    graph: fx.Graph = fwd_module.graph
+    nodes_list: list[fx.Node] = list(graph.nodes)
+    node_to_idx: dict[fx.Node, int] = {node: idx for idx, node in enumerate(nodes_list)}
```
did some timing checks and this is actually faster than the single pass approach, so sticking with this :). I think the `list()` function is efficient and better than an explicit loop.

xuanzhang816 committed:
Update on "activation offloading reordering for comp<>comm overlaps"

We introduce two flags for computation <> communication overlap:
* `activation_offload_sink_wait` -- with this turned on, the wait events in forward is sinked to the end of the graph so that the offload ops is not blocking
* `activation_reload_prefetch` -- with this turned on, the reload in the backward is prefetched just enough so that the reload operation is perfectly overlapped with computation.

Continue with the [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4) in the first PR in the stack:

4. AO on separate stream with reorders (i.e., additionally with `activation_offload_sink_wait = True` and `activation_reload_prefetch = True`)

- 7534 MB peak memory 
- 160.86 ms ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="869" height="123" alt="image" src="https://github.com/user-attachments/assets/6aa0c5df-5cc4-4711-8c1a-20f4a9d9145b" />






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

xuanzhang816 committed:
Update on "activation offloading reordering for comp<>comm overlaps"

We introduce two flags for computation <> communication overlap:
* `activation_offload_sink_wait` -- with this turned on, the wait events in forward is sinked to the end of the graph so that the offload ops is not blocking
* `activation_reload_prefetch` -- with this turned on, the reload in the backward is prefetched just enough so that the reload operation is perfectly overlapped with computation.

Continue with the [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4) in the first PR in the stack:

4. AO on separate stream with reorders (i.e., additionally with `activation_offload_sink_wait = True` and `activation_reload_prefetch = True`)

- 7534 MB peak memory 
- 160.86 ms ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="869" height="123" alt="image" src="https://github.com/user-attachments/assets/6aa0c5df-5cc4-4711-8c1a-20f4a9d9145b" />






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela chenyang78

[ghstack-poisoned]

xuanzhang816 commented:
@pytorchbot revert -m 'internal CI failure with lazy init on non GPU machines' -c ghfirst

xuanzhang816 committed:
Update on "activation offloading reordering for comp<>comm overlaps"

We introduce two flags for computation <> communication overlap:
* `activation_offload_sink_wait` -- with this turned on, the wait events in forward is sinked to the end of the graph so that the offload ops is not blocking
* `activation_reload_prefetch` -- with this turned on, the reload in the backward is prefetched just enough so that the reload operation is perfectly overlapped with computation.

Continue with the [example model](https://gist.github.com/xuanzhang816/ee2e3648123670f14ced9963858ee3b4) in the first PR in the stack:

4. AO on separate stream with reorders (i.e., additionally with `activation_offload_sink_wait = True` and `activation_reload_prefetch = True`)

- 7534 MB peak memory 
- 160.86 ms ms per-iteration runtime 
- trace for fwd and bwd as below (the "M" blocks are for Memcpy):
<img width="869" height="123" alt="image" src="https://github.com/user-attachments/assets/6aa0c5df-5cc4-4711-8c1a-20f4a9d9145b" />






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela chenyang78

[ghstack-poisoned]

xuanzhang816 commented:
@pytorchbot merge -f 'Minor update to skip tests when there is no GPU. Expecting all PR tests to pass'
